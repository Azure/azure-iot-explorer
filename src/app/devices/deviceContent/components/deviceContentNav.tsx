/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { Nav, INavLink, INavLinkGroup } from 'office-ui-fabric-react/lib/Nav';
import { Label } from 'office-ui-fabric-react/lib/Label';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { ROUTE_PARTS, ROUTE_PARAMS } from '../../../constants/routes';
import '../../../css/_deviceContentNav.scss';

export interface DeviceContentNavDataProps {
    deviceId: string;
    interfaceIds: string[];
    isLoading: boolean;
    isPnPDevice: boolean;
    selectedInterface: string;
    isEdgeDevice: boolean;
}

export interface DeviceContentNavDispatchProps {
    setInterfaceId: (interfaceId: string) => void;
}

interface DeviceContentNavState {
    expandedInterfaceMap: Map<string, boolean>;
}

export const NAV_LINK_ITEMS_PNP = [ROUTE_PARTS.INTERFACES, ROUTE_PARTS.SETTINGS, ROUTE_PARTS.PROPERTIES, ROUTE_PARTS.COMMANDS, ROUTE_PARTS.EVENTS];
export const NAV_LINK_ITEMS_NONPNP = [ROUTE_PARTS.IDENTITY, ROUTE_PARTS.TWIN, ROUTE_PARTS.EVENTS, ROUTE_PARTS.METHODS, ROUTE_PARTS.CLOUD_TO_DEVICE_MESSAGE];
export const NAV_LINK_ITEMS_NONPNP_NONEDGE = [ROUTE_PARTS.IDENTITY, ROUTE_PARTS.TWIN, ROUTE_PARTS.EVENTS, ROUTE_PARTS.METHODS, ROUTE_PARTS.CLOUD_TO_DEVICE_MESSAGE, ROUTE_PARTS.MODULE_IDENTITY];

export type DeviceContentNavProps = DeviceContentNavDataProps & DeviceContentNavDispatchProps & RouteComponentProps;
export default class DeviceContentNavComponent extends React.Component<DeviceContentNavProps, DeviceContentNavState> {
    constructor(props: DeviceContentNavProps) {
        super(props);
        const expandedInterfaceMap = new Map();
        if (this.props.selectedInterface) {
            expandedInterfaceMap.set(this.props.selectedInterface, true);
        }
        this.state = {expandedInterfaceMap};
    }

    public render(): JSX.Element {

        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <div className="view-scroll">
                        {this.createNavLinks(context)}
                    </div>
                )}
            </LocalizationContextConsumer>
        );
    }

    private readonly createNavLinks = (context: LocalizationContextInterface) => {
        const { deviceId, interfaceIds, isPnPDevice, isEdgeDevice } = this.props;
        const url = this.props.match.url;

        const navItems = isEdgeDevice ? NAV_LINK_ITEMS_NONPNP : NAV_LINK_ITEMS_NONPNP_NONEDGE;
        const nonPnpNavLinks = navItems.map((nav: string) => ({
            key: nav,
            name: context.t((ResourceKeys.deviceContent.navBar as any)[nav]), // tslint:disable-line:no-any
            url: `#${url}/${nav}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`
        }));

        const pnpNavGroupsLinks = isPnPDevice && interfaceIds && interfaceIds.map((id: string) => ({
            isExpanded: this.state.expandedInterfaceMap.get(id),
            links: NAV_LINK_ITEMS_PNP.map((nav: string): INavLink => ({
                key: `${id}-${nav}`,
                name: context.t((ResourceKeys.deviceContent.navBar as any)[nav]), // tslint:disable-line:no-any
                onClick: this.onNestedChildLinkClick,
                parentId: id,
                url: `#${url}/${ROUTE_PARTS.DIGITAL_TWINS}/${nav}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}&${ROUTE_PARAMS.INTERFACE_ID}=${id}`
            })),
            name: id,
            url: ''
        }));

        const groups = [];
        groups.push({
            links: nonPnpNavLinks,
            name: context.t(ResourceKeys.deviceContent.navBar.nonpnp),
        });
        if (isPnPDevice) {
            groups.push({
                links: pnpNavGroupsLinks,
                name: context.t(ResourceKeys.deviceContent.navBar.pnp),
            });
        }

        return (
            <div role="navigation">
                <Nav
                    onLinkExpandClick={this.onChildLinkExpand}
                    onRenderGroupHeader={this.onRenderGroupHeader}
                    groups={groups}
                />
            </div>
        );
    }

    private readonly onRenderGroupHeader = (group: INavLinkGroup) => {
        return <Label className="nav-label">{group.name}</Label>;
      }

    private readonly onNestedChildLinkClick = (ev?: React.MouseEvent<HTMLElement>, item?: INavLink) => {
        this.props.setInterfaceId(item.parentId);
    }

    private readonly onChildLinkExpand = (ev?: React.MouseEvent<HTMLElement>, item?: INavLink) => {
        const expandedInterfaceMap = new Map(this.state.expandedInterfaceMap);
        expandedInterfaceMap.set(item.name, !item.isExpanded);
        this.setState({
            expandedInterfaceMap
        });
    }
}
