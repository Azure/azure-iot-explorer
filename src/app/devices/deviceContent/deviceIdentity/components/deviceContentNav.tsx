/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useLocation, useRouteMatch } from 'react-router-dom';
import { Nav, INavLinkGroup } from 'office-ui-fabric-react/lib/Nav';
import { useLocalizationContext } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { ROUTE_PARTS, ROUTE_PARAMS } from '../../../../constants/routes';
import { getDeviceIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import '../../../../css/_deviceContentNav.scss';

export interface DeviceContentNavDataProps {
    isEdgeDevice: boolean;
}

export const NAV_LINK_ITEMS = [ROUTE_PARTS.IDENTITY, ROUTE_PARTS.TWIN, ROUTE_PARTS.EVENTS, ROUTE_PARTS.METHODS, ROUTE_PARTS.CLOUD_TO_DEVICE_MESSAGE];
export const NAV_LINK_ITEMS_NONEDGE = [...NAV_LINK_ITEMS, ROUTE_PARTS.MODULE_IDENTITY, ROUTE_PARTS.DIGITAL_TWINS];

export type DeviceContentNavProps = DeviceContentNavDataProps;
export const DeviceContentNavComponent: React.FC<DeviceContentNavProps> =  (props: DeviceContentNavProps) => {
    const { t } = useLocalizationContext();
    const { isEdgeDevice } = props;
    const {  search } = useLocation();
    const { url } = useRouteMatch();
    const deviceId = getDeviceIdFromQueryString(search);

    const navItems = isEdgeDevice ? NAV_LINK_ITEMS : NAV_LINK_ITEMS_NONEDGE;
    const navLinks = navItems.map((nav: string) => ({
        key: nav,
        name: t((ResourceKeys.deviceContent.navBar as any)[nav]), // tslint:disable-line:no-any
        url: `#${url}/${nav}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`
    }));

    const groups: INavLinkGroup[] = [{ links: navLinks }];

    return (
        <div role="navigation">
            <Nav groups={groups}/>
        </div>
    );
};
