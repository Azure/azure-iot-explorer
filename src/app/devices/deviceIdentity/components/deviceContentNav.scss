/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
 @import '../../../css/themes';

.nav-link-left {
    a:link, a:visited, a:active {
        text-decoration: none;
        @include themify($themes) {
            color: themed('menuLinkColor');
        }
    }

    a:hover {
        @include themify($themes) {
            color: themed('menuLinkColor_Hover');
            background-color: themed('menuLinkBackground_Hover');
        }
    }
}
