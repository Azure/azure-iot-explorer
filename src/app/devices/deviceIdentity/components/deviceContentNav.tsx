/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useRouteMatch } from 'react-router-dom';
import { Nav, INavLinkGroup, INavLink, Announced } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { ROUTE_PARTS, ROUTE_PARAMS } from '../../../constants/routes';
import { NAVIGATE_BACK } from '../../../constants/iconNames';
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../../shared/utils/queryStringHelper';
import './deviceContentNav.scss';

export const NAV_LINK_ITEMS_DEVICE = [ROUTE_PARTS.IDENTITY, ROUTE_PARTS.TWIN, ROUTE_PARTS.EVENTS, ROUTE_PARTS.METHODS, ROUTE_PARTS.CLOUD_TO_DEVICE_MESSAGE, ROUTE_PARTS.MODULE_IDENTITY];
export const NAV_LINK_ITEMS_NONEDGE_DEVICE = [...NAV_LINK_ITEMS_DEVICE, ROUTE_PARTS.DIGITAL_TWINS];
export const NAV_LINK_ITEMS_MODULE = [ROUTE_PARTS.MODULE_DETAIL, ROUTE_PARTS.MODULE_TWIN, ROUTE_PARTS.MODULE_METHOD, ROUTE_PARTS.MODULE_EVENTS, ROUTE_PARTS.MODULE_PNP];
// tslint:disable-next-line: no-any
const navIcons = {} as any;
navIcons[ROUTE_PARTS.IDENTITY] = 'Server';
navIcons[ROUTE_PARTS.TWIN] = 'ReopenPages';
navIcons[ROUTE_PARTS.EVENTS] = 'Message';
navIcons[ROUTE_PARTS.METHODS] = 'Remote';
navIcons[ROUTE_PARTS.CLOUD_TO_DEVICE_MESSAGE] = 'Mail';
navIcons[ROUTE_PARTS.MODULE_IDENTITY] = 'TFVCLogo';
navIcons[ROUTE_PARTS.DIGITAL_TWINS] = 'PlugDisconnected';
navIcons[ROUTE_PARTS.MODULE_DETAIL] = 'TFVCLogo';
navIcons[ROUTE_PARTS.MODULE_TWIN] = 'ReopenPages';
navIcons[ROUTE_PARTS.MODULE_METHOD] = 'Remote';
navIcons[ROUTE_PARTS.MODULE_EVENTS] = 'Message';
navIcons[ROUTE_PARTS.MODULE_PNP] = 'PlugDisconnected';

export interface DeviceContentNavProps {
    isEdgeDevice: boolean;
}

export const DeviceContentNavComponent: React.FC<DeviceContentNavProps> =  (props: DeviceContentNavProps) => {
    const { t } = useTranslation();
    const { isEdgeDevice } = props;
    const { search, pathname } = useLocation();
    const { url } = useRouteMatch();
    const [ selectedRoute, setSelectedRoute] = React.useState<string>();
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);

    const navItems = moduleId ? NAV_LINK_ITEMS_MODULE : (isEdgeDevice ? NAV_LINK_ITEMS_DEVICE : NAV_LINK_ITEMS_NONEDGE_DEVICE);
    const navLinks: INavLink[] = [];

    if (moduleId) {
        navLinks.push({
            iconProps: {iconName: NAVIGATE_BACK},
            key: NAVIGATE_BACK,
            name: t(ResourceKeys.moduleIdentity.detail.command.back),
            url: `#${url}/${ROUTE_PARTS.MODULE_IDENTITY}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`
        });
        navItems.forEach((nav: string) => {
            const navUrl = `${url}/${ROUTE_PARTS.MODULE_IDENTITY}/${nav}/`;
            navLinks.push({
                iconProps: { iconName: navIcons[nav] },
                key: nav,
                name: t((ResourceKeys.breadcrumb as any)[nav]), // tslint:disable-line:no-any
                url: `#${navUrl}?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}&${ROUTE_PARAMS.MODULE_ID}=${encodeURIComponent(moduleId)}`
            });
        });
    }

    else {
        navItems.forEach((nav: string) => {
            const navUrl = `${url}/${nav}`;
            navLinks.push({
                iconProps: { iconName: navIcons[nav] },
                key: nav,
                name: t((ResourceKeys.breadcrumb as any)[nav]), // tslint:disable-line:no-any
                url: `#${navUrl}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`
            });
        });
    }

    const groups: INavLinkGroup[] = [{ links: navLinks }];

    React.useEffect(() => {
        const foundRoutes = navItems.filter(nav => pathname.includes(nav));
        const currentRoute = foundRoutes?.[foundRoutes.length - 1] ?? ''; // because PnP has its own 'events' route.
        setSelectedRoute(currentRoute);
    }, [pathname]); // tslint:disable-line: align

    return (
        <div role="navigation" className="nav-link-left">
            <Nav groups={groups} selectedKey={selectedRoute}/>
            {selectedRoute && <Announced message={`${t(ResourceKeys.breadcrumb.navigate)} ${selectedRoute}`}/>}
        </div>
    );
};
