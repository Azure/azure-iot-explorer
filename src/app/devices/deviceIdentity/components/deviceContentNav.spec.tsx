/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { Nav } from '@fluentui/react';
import { DeviceContentNavComponent, DeviceContentNavProps, NAV_LINK_ITEMS_DEVICE, NAV_LINK_ITEMS_NONEDGE_DEVICE } from './deviceContentNav';

jest.mock('react-router-dom', () => ({
    useLocation: () => ({ search: '?deviceId=test', pathname: '' }),
    useRouteMatch: () => ({ url: '' })
}));

describe('deviceContentNav', () => {
    const getComponent = (overrides = {}) => {
        const navDataProps: DeviceContentNavProps = {
            isEdgeDevice: true
        };

        const props = {
            ...navDataProps,
            ...overrides,
        };

        return <DeviceContentNavComponent {...props} />;
    };

    it('matches snapshot when the device is edge', () => {
        expect(shallow(getComponent())).toMatchSnapshot();
        const wrapper = mount(getComponent());
        const navigation = wrapper.find(Nav);
        expect(navigation.props().groups[0].links.length).toEqual(NAV_LINK_ITEMS_DEVICE.length);
    });

    it('matches snapshot when the device is not edge', () => {
        expect(shallow(getComponent({ isEdgeDevice: false }))).toMatchSnapshot();
        const wrapper = mount(getComponent({ isEdgeDevice: false }));
        const navigation = wrapper.find(Nav);
        expect(navigation.props().groups[0].links.length).toEqual(NAV_LINK_ITEMS_NONEDGE_DEVICE.length);
    });

    it('shows non-pnp nav when component is loading', () => {
        const wrapper = mount(getComponent({isLoading: true}));

        const navigation = wrapper.find(Nav);
        expect(navigation.props().groups[0].links.length).toEqual(NAV_LINK_ITEMS_DEVICE.length);
    });
});
