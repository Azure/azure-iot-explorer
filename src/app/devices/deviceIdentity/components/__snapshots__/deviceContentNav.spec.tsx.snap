// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deviceContentNav matches snapshot when the device is edge 1`] = `
<div
  className="nav-link-left"
  role="navigation"
>
  <StyledNavBase
    groups={
      Array [
        Object {
          "links": Array [
            Object {
              "iconProps": Object {
                "iconName": "Server",
              },
              "key": "identity",
              "name": "breadcrumb.identity",
              "url": "#/identity/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "ReopenPages",
              },
              "key": "twin",
              "name": "breadcrumb.twin",
              "url": "#/twin/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "Message",
              },
              "key": "events",
              "name": "breadcrumb.events",
              "url": "#/events/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "Remote",
              },
              "key": "methods",
              "name": "breadcrumb.methods",
              "url": "#/methods/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "Mail",
              },
              "key": "cloudToDeviceMessage",
              "name": "breadcrumb.cloudToDeviceMessage",
              "url": "#/cloudToDeviceMessage/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "TFVCLogo",
              },
              "key": "moduleIdentity",
              "name": "breadcrumb.moduleIdentity",
              "url": "#/moduleIdentity/?deviceId=test",
            },
          ],
        },
      ]
    }
  />
</div>
`;

exports[`deviceContentNav matches snapshot when the device is not edge 1`] = `
<div
  className="nav-link-left"
  role="navigation"
>
  <StyledNavBase
    groups={
      Array [
        Object {
          "links": Array [
            Object {
              "iconProps": Object {
                "iconName": "Server",
              },
              "key": "identity",
              "name": "breadcrumb.identity",
              "url": "#/identity/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "ReopenPages",
              },
              "key": "twin",
              "name": "breadcrumb.twin",
              "url": "#/twin/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "Message",
              },
              "key": "events",
              "name": "breadcrumb.events",
              "url": "#/events/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "Remote",
              },
              "key": "methods",
              "name": "breadcrumb.methods",
              "url": "#/methods/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "Mail",
              },
              "key": "cloudToDeviceMessage",
              "name": "breadcrumb.cloudToDeviceMessage",
              "url": "#/cloudToDeviceMessage/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "TFVCLogo",
              },
              "key": "moduleIdentity",
              "name": "breadcrumb.moduleIdentity",
              "url": "#/moduleIdentity/?deviceId=test",
            },
            Object {
              "iconProps": Object {
                "iconName": "PlugDisconnected",
              },
              "key": "ioTPlugAndPlay",
              "name": "breadcrumb.ioTPlugAndPlay",
              "url": "#/ioTPlugAndPlay/?deviceId=test",
            },
          ],
        },
      ]
    }
  />
</div>
`;
